Elasticsearch: Create Raw Data metric to render raw JSON docs in columns in the new table panel

ivanahuckova opened the pull request:
**What this PR does / why we need it**:

Follows up on https://github.com/grafana/grafana/pull/24521. To be backwards compatible, I have created Raw Document v2 metric to process results as DataFrames and not SeriesList.

Blocker: 
When switching between the Raw Document and Raw Document v2 table does not update as new request is not sent (because the request is the same).

**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/grafana/issues/24163

ivanahuckova committed:
test

ivanahuckova committed:
WIP: Create v2 version

ivanahuckova committed:
Update tests, remove conosole logs, refactor

ivanahuckova committed:
Remove incorrect types

ivanahuckova commented:
@marefr @torkelo Could you please check this out and share your thoughts? I have followed up on https://github.com/grafana/grafana/pull/24521 and created a new metric as suggested. 

Also, do you have any suggestions on how to fix the blocker - when switching between the Raw Document and Raw Document v2 table does not update as new request is not sent (because the request is the same).

ivanahuckova committed:
Update type

torkelo commented:
not sure I understand why a new query is not issued, you will need to trigger a new query for the result processing to be re-executed 

 `Raw Document v2` is not great. Maybe

* Document fields
* Raw data

and then change text label (not value) for the old Raw document to Raw document (legacy).

ivanahuckova committed:
Rename legacy and new metrics

ivanahuckova committed:
Update

ivanahuckova committed:
Run request when Raw Data tto Raw Document switch

ivanahuckova commented:
> not sure I understand why a new query is not issued, you will need to trigger a new query for the result processing to be re-executed

Figured that out  in query_ctrl we are checking if queries are the same. I have added exception for Raw Data and Raw Doocument.

> Raw Document v2 is not great. Maybe
Document fields
Raw data
and then change text label (not value) for the old Raw document to Raw document (legacy).

Changed to Raw Data, as I feel it is more clear and for intuitive for users based on Raw Document naming. Also they are right next to each other, so users can easily try it out, when they see Raw document (legacy).

marefr reviewed (changes requested):
Have tested this. In general I wonder whether introducing a different Metric is the way to go. Worth consider having an option above or below `Size` when Raw Document is selected. Maybe a switch/boolean for legacy format that's true if this property is null/undefined and when adding a new query and selecting Raw Document this gets set to false per default. Doing this gives us possibility to also add some tooltip/description of what it means. Thought?  

Found some bugs. 

Using Raw Data I cannot edit size option: 
![image](https://user-images.githubusercontent.com/1668778/87302316-78cc6b00-c511-11ea-9f5f-f0a5e8b4d46e.png)

Comparing with old table panel and Raw Document query it doesn't include _source as a column, but using Raw Data does which feels wrong:
![image](https://user-images.githubusercontent.com/1668778/87302361-8c77d180-c511-11ea-8c6d-0854a3d302a1.png)

ivanahuckova committed:
Fix size updating

ivanahuckova committed:
Remove _source field from table results as we are showing each source field as column

ivanahuckova commented:
>Using Raw Data I cannot edit size option

✅ Nice catch! Fixed in https://github.com/grafana/grafana/pull/26233/commits/2096cd032514b22c7ac4ae16fe6ceedd7a3dc3e0 

>Comparing with old table panel and Raw Document query it doesn't include _source as a column, but using Raw Data does which feels wrong

✅ Removed in https://github.com/grafana/grafana/pull/26233/commits/31a9d5f81b79b91a12a7e0b74f172ff8afc4bdd3

>Have tested this. In general I wonder whether introducing a different Metric is the way to go. Worth consider having an option above or below Size when Raw Document is selected. 

Well yeah, that is a really good question and I think I might like the solution that you have just proposed slightly better. I have created a new metric as both you and @torkelo have suggested it in https://github.com/grafana/grafana/pull/24521#issuecomment-656009423 and https://github.com/grafana/grafana/issues/24163#issuecomment-642541344, so I thought that there was a consensus that this would be the best option. But I try to create another PR with the other solution and we can just choose whichever we like more. But we should make sure that this fix goes to 7.1 as it was promised. 🙂

marefr commented:
Sounds good

ivanahuckova committed:
Remove _source just for metrics, not logs

torkelo commented:
> Maybe a switch/boolean for a legacy format that's true if this property is null/undefined and when adding a new query and selecting Raw Document this gets set to false per default. Doing this gives us the possibility to also add some tooltip/description of what it means. Thought? 

I generally like this idea better than the new "Raw data". The only problem is that we would like the default to be the new format (so legacy should default to false), but if undefined (ie not in the in the persisted model) this would mean it defaulted to the old model? to introduce this property without a migration we would need it inverted "useNewFormat" and have it always be true for new models. and live with it for a while. Maybe this is ok, we can remove it in a year or so

ivanahuckova commented:
I have created anew  PR for the "switch-idea" https://github.com/grafana/grafana/pull/26323.

marefr committed:
Merge remote-tracking branch 'origin/master' into ivana/elastic-table

ivanahuckova committed:
Revert "Remove _source just for metrics, not logs"

This reverts commit 611b6922f762afa0e76fd8679a9b8160bca74e6a.

ivanahuckova committed:
Revert "Remove _source field from table results as we are showing each source field as column"

This reverts commit 31a9d5f81b79b91a12a7e0b74f172ff8afc4bdd3.

marefr reviewed (dismissed):
Logs in Explore seems to be broken (works in master):
![image](https://user-images.githubusercontent.com/1668778/87526548-43e42380-c68b-11ea-9ca8-fedf47b04981.png)

If this was caused by removing _source column I'm fine with skipping that change if it resolves explore/logs problem

ivanahuckova committed:
Add vis preference for logs

ivanahuckova committed:
Update visualisation to logs

ivanahuckova committed:
Merge branch 'ivana/elastic-table' of https://github.com/grafana/grafana into ivana/elastic-table

ivanahuckova commented:
@marefr fixed in https://github.com/grafana/grafana/pull/26233/commits/3687e945e34698e644d95460bf3697b6433cbcc0

ivanahuckova commented:
![Kapture 2020-07-15 at 11 37 03](https://user-images.githubusercontent.com/30407135/87529633-90c9f900-c68f-11ea-8c8e-a50eca4520d5.gif)

ivanahuckova committed:
Revert "Revert "Remove _source just for metrics""

This reverts commit a102ab2894e7a9ca6eee307aa9b60dfea169c718.

ivanahuckova commented on public/app/plugins/datasource/elasticsearch/query_ctrl.ts:
```diff
@@ -51,9 +51,15 @@ export class ElasticQueryCtrl extends QueryCtrl {
 
   queryUpdated() {
+    // As Raw Data and Raw Document have the same request, we need to run refresh if they are updated
+    const isPossiblyRawDataSwitch = this.target.metrics.some(
```
Will create another PR for this.
